api/src/main/java/org/apache/brooklyn/api/internal/AbstractBrooklynObjectSpec.java

```diff
@@ -187,7 +202,24 @@ public final String getDisplayName() {
     }
     
     public final String getCatalogItemId() {
-        return catalogItemId;
+        if (catalogItemIdStack.size() != 0) {
+            return catalogItemIdStack.getFirst();
+        }
+        return null;
```

Comment:
Seems preferable not to retain catalogItemId if we can get away without it, see
comment https://github.com/apache/brooklyn-server/pull/338#discussion_r102679601 above.